### PR TITLE
Requests fully implemented

### PR DESCRIPTION
**Pull request by Mackey22**
We should probably go over this together. Grab me when you have time tomorrow.

### CODE REVIEW

File: hca/define_api.py [hunk 1]
```diff
@@ -2,36 +2,133 @@
...
 
-    def get_spec(self):
+        self.positional_args = {endpoint: [arg['argument'] for arg in self.param_data[endpoint]['positional']] for endpoint in self.param_data}
+
+    def get_spec(self, test):
         """Load the API specification."""
-        with open(os.path.dirname(os.path.realpath(__file__)) + "/api_spec.json") as fp:
+        url = os.path.dirname(os.path.realpath(__file__)) + "/api_spec.json"
+        if test:
+            url = os.path.dirname(os.path.realpath(__file__)) + "/../test/test.json"
```
Comment (Mackey22) [inline hca/define_api.py]: Realized I'd never responded to these threads, but I've done each of the changes you recommended. In this case, api_spec is the one downloaded from the remote api whereas test.json is a spec that I've been using to test different features myself.

File: hca/define_api.py [hunk 2]
```diff
@@ -2,36 +2,133 @@
...
         """Initialize the CLI API."""
-        self.spec = self.get_spec()
-        self.base_url = self.spec['host'] + self.spec['basePath']
-        self.parser = get_parser(self.spec)
+        spec = self.get_spec(test)
+        self.base_url = "https://" + spec['host'] + spec['basePath']
+        self.parser, self.param_data = get_parser(spec)
 
-    def get_spec(self):
+        self.positional_args = {endpoint: [arg['argument'] for arg in self.param_data[endpoint]['positional']] for endpoint in self.param_data}
```
Comment (Mackey22) [inline hca/define_api.py]: 👍

File: hca/define_api.py [hunk 3]
```diff
@@ -2,36 +2,133 @@
...
+import pprint
 import requests
 from .parser import get_parser
 
 
 class API:
     """Class for interacting with the REST Api."""
 
-    def __init__(self, api_url, user):
+    def __init__(self, user, password, test=False):
```
Comment (Mackey22) [inline hca/define_api.py]: Yeah that makes much more sense. Wasn't sure how we were doing authentication at the beginning - I think that user, pw combo was from back then. Not sure why it got re-added.

File: hca/full_upload.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+import sys, os, json
+from io import open
+import argparse
+from datetime import datetime
+import uuid
+import pprint
+
+from .constants import Constants
+from .upload_to_s3 import upload_to_s3
```
Comment (ttung) [inline hca/full_upload.py]: what is your plan for making this multi-cloud?

**Comment by ttung** [comment]
Please consider splitting into separate diffs.

File: .gitmodules [hunk 5]
```diff
@@ -0,0 +1,3 @@
+[submodule "checksum_reader"]
```
Comment (Mackey22) [inline .gitmodules]: Not in pypi at this point. Would be great to get that up there actually this whole submodule thing is annoying. @sampierson I can look at that with you tomorrow if you have time.

File: hca/full_upload.py (hunk 4 above)
Comment (Mackey22) [inline hca/full_upload.py]: I'll rename file and class to `upload_to_cloud`.

File: hca/define_api.py [hunk 6]
```diff
@@ -1,37 +1,165 @@
...
+    def __init__(self, test=False):
         """Initialize the CLI API."""
-        self.spec = self.get_spec()
-        self.base_url = self.spec['host'] + self.spec['basePath']
-        self.parser = get_parser(self.spec)
+        spec = self.get_spec(test)
+        self.base_url = "https://" + spec['host'] + spec['basePath']
+        self.parser, self.param_data = get_parser(spec)
+
+    def get_spec(self, test):
```
Comment (Mackey22) [inline hca/define_api.py]: Will default to False

File: hca/define_api.py [hunk 7]
```diff
@@ -1,37 +1,165 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
-import json, sys, os
+import json, sys, os, ast
```
Comment (Mackey22) [inline hca/define_api.py]: Will do. Should I not do one import per line if coming from same module? Also seen standards where w/in package imports are below other imports. Is this something we do as well?

**Comment by Mackey22** [comment]
@ttung will do. Focusing on download for demo tomorrow first then will try to clean this up.

File: hca/define_api.py (hunk 7 above)
Comment (ttung) [inline hca/define_api.py]: Yeah, if it's from the same module, you can do multiline.

**Comment by Mackey22** [comment]
Refactored into smaller pull requests.

**Pull request closed without merging**